lib/print/serial.c

```diff
@@ -28,87 +28,143 @@
-#error "LIB_PCI is not defined"
-#endif
-#else
-	return (uint64_t)SERIAL_BASE;
+#ifndef SERIAL_BAUD
+#define SERIAL_BAUD 115200 //default baud
 #endif
+	unsigned char c;
+	unsigned div = 115200/SERIAL_BAUD;
+
+	if (serial_base == -1ULL)
+		return;
+
+	g_ser_dev.set(serial_base, UART_REG_LCR, 0x3); // 8n1
+	g_ser_dev.set(serial_base, UART_REG_IER, 0x0); // disable interrupt
+	g_ser_dev.set(serial_base, UART_REG_FCR, 0x0); // disable fifo
+	g_ser_dev.set(serial_base, UART_REG_MCR, 0x3); // DTR + RTS
+
+	c = g_ser_dev.get(serial_base, UART_REG_LCR);
+	g_ser_dev.set(serial_base, UART_REG_LCR,  c | BIT(8));          // Set DLAB(Divisor Latch Access Bit)
+	g_ser_dev.set(serial_base, UART_REG_DLL,  div & 0xff);          // Set Divisor Latch Low byte
+	g_ser_dev.set(serial_base, UART_REG_DLM,  (div >> 8) & 0xff);   // Set Divisor Latch High byte
+	g_ser_dev.set(serial_base, UART_REG_LCR,  c & ~BIT(8));         // Clear DLAB
 }
 
-static UNUSED void serial_pci_mmio_init(uint64_t serial_base)
+#ifdef SERIAL_PCI
+void serial_init(boolean_t setup)
 {
-	// temp workaround: get UART out of reset
-	serial_set_reg(serial_base, 0x204, 0x00);
-	serial_set_reg(serial_base, 0x204, 0x07);
-	// update and enable (M, N) clock divider.  Magic register value is from PythonSv.
-	*(uint32_t *)(serial_base + 0x200) = 0xffff04b5;
-
-	serial_set_reg(serial_base, 0x0c, 0x83);
-	serial_set_reg(serial_base, 0x00, 0x01);
-	serial_set_reg(serial_base, 0x04, 0x00);
-	serial_set_reg(serial_base, 0x0c, 0x03);
-
-	// enable & reset (receive and transmit, 64 byte) FIFO:
-	serial_set_reg(serial_base, 0x08, 0x27);
-
-	// enable logging & printing of console output.
-	serial_set_reg(serial_base, 0X1C, 0x00);
-}
+	uint32_t bar0;
+	uint16_t dev;
+	uint16_t class;
+
+	dev = SERIAL_DEV;
+	class = pci_read16(dev, PCI_CLASS_REG);
+	if ((class != PCI_CLASS_COM_SERIAL) &&
+	    (class != PCI_CLASS_COM_MODEM)) {
+		return;
+	}
+
+	bar0 = pci_read32(dev, PCI_BAR_REG(0));
+
+	if (PCI_BAR_IS_IO_SPACE(bar0)) {
+		pci_write16(dev, PCI_CMD_REG, pci_read16(dev, PCI_CMD_REG) | PCI_CMD_IO_SPACE);
+		g_ser_dev.base = bar0 & PCI_BAR_IO_MASK;
+		g_ser_dev.get = serial_io_get;
+		g_ser_dev.set = serial_io_set;
+	} else {
+		pci_write16(dev, PCI_CMD_REG, pci_read16(dev, PCI_CMD_REG) | PCI_CMD_MEMORY_SPACE);
+		g_ser_dev.base = bar0 & PCI_BAR_MMIO_MASK;
+		g_ser_dev.get = serial_mmio_get;
+		g_ser_dev.set = serial_mmio_set;
+	}
 
-void serial_init(UNUSED uint64_t serial_base)
+	if (setup) {
+		serial_8250_init(g_ser_dev.base);
+	}
+
+}
+#else
+void serial_init(boolean_t setup)
 {
-#if (defined SERIAL_PCI) && (defined SERIAL_MMIO)
-	serial_pci_mmio_init(serial_base);
+	g_ser_dev.base = SERIAL_BASE;
+
+#ifdef SERIAL_IO
+	g_ser_dev.get = serial_io_get;
+	g_ser_dev.set = serial_io_set;
+#elif defined SERIAL_MMIO
+	g_ser_dev.get = serial_mmio_get;
+	g_ser_dev.set = serial_mmio_set;
+#else
+#error "Serial type(SERIAL_IO/SERIAL_MMIO) is not defined"
 #endif
-}
 
+	if (setup) {
+		serial_8250_init(g_ser_dev.base);
+	}
```

Comment:
merge it with SERIAL_PCI above. how about add a "init_serial_dev()" to setup g_ser_dev for different conditions and in serial_init() to call it.